gpjax/kernels/non_euclidean/categorical.py

```diff
@@ -0,0 +1,126 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+
+from dataclasses import dataclass
+from typing import NamedTuple
+import jax.numpy as jnp
+from jaxtyping import Float
+import tensorflow_probability.substrates.jax as tfp
+
+from gpjax.base import (
+    param_field,
+    static_field,
+)
+from gpjax.kernels.base import AbstractKernel
+
+from gpjax.typing import (
+    Array,
+    ScalarInt,
+)
+
+tfb = tfp.bijectors
+
+CatKernelParams = NamedTuple(
+    "DictKernelParams",
+    [("sdev", Float[Array, "N 1"]), ("cholesky_lower", Float[Array, " N*(N-1)//2"])],
+)
+
+
+@dataclass
+class CatKernel(AbstractKernel):
+    r"""The categorical kernel is defined for a fixed number of values of categorical input.
+
+    It stores a standard dev for each input value (i.e. the diagonal of the gram), and a lower cholesky factor for correlations.
+    It returns the corresponding values from an the gram matrix when called.
+
+    Args:
+        sdev (Float[Array, "N"]): The standard deviation parameters, one for each input space value.
+        cholesky_lower (Float[Array, "N*(N-1)//2 N"]): The parameters for the Cholesky factor of the gram matrix.
+        inspace_vals (list): The values in the input space this DictKernel works for. Stored for order reference, making clear the indices used for each input space value.
+
+    Raises:
+        ValueError: If the number of diagonal variance parameters does not match the number of input space values.
+    """
+
+    sdev: Float[Array, " N"] = param_field(jnp.ones((2,)), bijector=tfb.Softplus())
+    cholesky_lower: Float[Array, "N N"] = param_field(
+        jnp.eye(2), bijector=tfb.CorrelationCholesky()
+    )
+    inspace_vals: list = static_field(None)
+    name: str = "Dictionary Kernel"
+    input_1hot: bool = static_field(False)
+
+    def __post_init__(self):
+        if self.inspace_vals is not None and len(self.inspace_vals) != len(self.sdev):
+            raise ValueError(
+                f"The number of sdev parameters ({len(self.sdev)}) has to match the number of input space values ({len(self.inspace_vals)}), unless inspace_vals is None."
+            )
+
+    @property
+    def explicit_gram(self):
+        L = self.sdev.reshape(-1, 1) * self.cholesky_lower
+        return L @ L.T
+
+    def __call__(  # TODO not consistent with general kernel interface
```

Comment:
This is True of the GraphKernel too. Not helpful, I know, but maybe there's an alternative abstraction that is more appropriate for non-Euclidean kernels.

Reply:
The this is because `AbstractKernel.__call__(x, y)` requires float array for x and y. Insteadm maybe the right signature for this baseclass would be
`def __call__(self, x: Num[Array, " D"], y: Num[Array, " D"]) -> ScalarFloat:`
because then the categorical kernel could specialize to ScalarInt if I'm not mistaken.

Reply:
Yes. I agree this would be the more general signature.